feat: Add a package to support Dio request tracking

fuzzybinary opened the pull request:
### What and why?

The `datadog_dio` package adds a Dio interceptor for supporting tracing of Dio requests. This is to support Dio HTTP Client Adapters that don't use `dart:io` or the native wrapper packages

This package uses a Dio `Interceptor` over an `HttpClientAdapter`. `Interceptor`s seem like the correct / less intrusive way of tracking requests from Dio, and although they can be skipped. inserting the interceptor into the first slot should avoid issues with clients accidentally skipping the interception process, as they are processed FIFO.

refs: RUM-12376

### Review checklist

- [ ] This pull request has appropriate unit and / or integration tests 
- [ ] This pull request references a Github or JIRA issue

fuzzybinary committed:
wip: Initial commit for dio package

fuzzybinary committed:
wip: Add Dio interceptor and integration tests

Still needs unit tests but all Dio interception functionality appears to be working.

fuzzybinary committed:
feat: Add a Dio package

The `datadog_dio` package adds a Dio interceptor for supporting tracing of Dio requests. This is to support Dio HTTP Client Adapters that don't use `dart:io` or the native wrapper packages

refs: RUM-12376

0xnm commented on .vscode/launch.json:
```diff
@@ -25,7 +25,7 @@
         {
             "name": "tracking http client example",
             "type": "dart",
-            "cwd": "packages/datadog_tracking_http_client/example",
```
This package/example still exists though. Probably it should be a dedicated block for `packages/datadog_dio/example` instead of changing existing one?

fuzzybinary replied:
I should at least revert this yes. Having this may run configurations is kind of a pain though :)

0xnm commented on packages/datadog_dio/README.md:
```diff
@@ -0,0 +1,17 @@
+
+# Datadog Dio integration
+
+> A package for use with Dio and the Datadog SDK, used tro track performance of HTTP calls and enable Datadog Distributed Tracing.
```
```suggestion
> A package for use with Dio and the Datadog SDK, used to track performance of HTTP calls and enable Datadog Distributed Tracing.
```

0xnm commented on packages/datadog_dio/example/README.md:
Looks like this file should either be removed or updated with the actual project description.

0xnm commented on packages/datadog_dio/example/android/app/build.gradle.kts:
```diff
@@ -0,0 +1,44 @@
+}
+
+android {
+    namespace = "com.example.example"
```
This has to be updated? Although the `kotlin/com/example/example` path should be updated as well.

0xnm commented on packages/datadog_dio/example/android/app/build.gradle.kts:
```diff
@@ -0,0 +1,44 @@
+
+    defaultConfig {
+        // TODO: Specify your own unique Application ID (https://developer.android.com/studio/build/application-id.html).
+        applicationId = "com.example.example"
```
this has to be updated as well

0xnm commented on packages/datadog_dio/example/android/app/src/profile/AndroidManifest.xml:
there is no `profile` flavour declared in the `build.gradle`, so I think this file and folder are not needed.

fuzzybinary replied:
It's defined by Flutter's `build.gradle`

0xnm commented on packages/datadog_dio/example/lib/scenario_config.dart:
```diff
@@ -0,0 +1,32 @@
+class RumAutoInstrumentationScenarioConfig {
+  final List<String> firstPartyHosts;
+  final String firstPartyGetUrl;
+  final String? firstPartyPostUrl;
```
it seems this one is never actually set?

fuzzybinary replied:
It is. It's set by the integration tests.

0xnm commented on packages/datadog_dio/example/web/manifest.json:
```diff
@@ -0,0 +1,35 @@
+    "display": "standalone",
+    "background_color": "#0175C2",
+    "theme_color": "#0175C2",
+    "description": "A new Flutter project.",
```
this needs to be updated I think

0xnm commented on packages/datadog_dio/lib/datadog_dio.dart:
```diff
@@ -0,0 +1,37 @@
+/// A set of callbacks that allow you to provide attributes that should be
+/// attached to a Datadog RUM resource created from a [DatadogDioInterceptor]. This
+/// callback is called as part of the [Interceptor.onResponse] and [Interceptor.onError]
+/// callbacks..
```
```suggestion
/// callbacks.
```

0xnm commented on packages/datadog_dio/test/datadog_dio_interceptor_test.dart:
```diff
@@ -0,0 +1,612 @@
+    when(() => mockDatadog
+            .headerTypesForHost(any(that: HasHost(equals('test_url')))))
+        .thenReturn({TracingHeaderType.datadog});
+    //when(() => mockDatadog.platform).thenReturn(mockPlatform);
```
should it be left like this?

fuzzybinary replied:
No it should have been removed.

fuzzybinary committed:
Fix web integration tests.

fuzzybinary committed:
chore: Dio interceptor documentation

fuzzybinary committed:
Apply suggestions from code review

Co-authored-by: Nikita Ogorodnikov <4046447+0xnm@users.noreply.github.com>

fuzzybinary committed:
Update packages/datadog_dio/lib/datadog_dio.dart

Co-authored-by: Nikita Ogorodnikov <4046447+0xnm@users.noreply.github.com>

brett0000FF commented on packages/datadog_dio/README.md:
```diff
@@ -0,0 +1,37 @@
+
+## Use with other Dio interceptors
+
+`addDatadogInterceptor` adds the Datadog interceptor as the first interceptor in your list of interceptors, which is important to ensure that all network requests from Dio are sent to Datadog, as other interceptors may decide not to forward information down the interceptor chain. For this reason, it is important to make sure you call `addDatadogInterceptor` actor any other configuration of Dio is ...
```
```suggestion
`addDatadogInterceptor` adds the Datadog interceptor as the first interceptor in your list of interceptors, which is important to ensure that all network requests from Dio are sent to Datadog, as other interceptors may decide not to forward information down the interceptor chain. For this reason, it is important to make sure you call `addDatadogInterceptor` after any other configuration of Dio is complete.
```

brett0000FF commented on packages/datadog_dio/README.md:
```diff
@@ -0,0 +1,37 @@
+
+Clients that want to track all network requests, including those made by `dart:io` and widgets like `NetworkImage` can continue to use [`datadog_tracking_http_client`](https://pub.dev/packages/datadog_tracking_http_client) to capture these requests. However, depending on your setup, the global overide method used in `enableHttpTracking` may cause resources to be double reported (once by the globa ...
+
+To avoid this, we recommend using `ignoreUrlPatterns` parameter when calling `enableHttpTracking` to ignore ignore requests made by your Dio client.
```
```suggestion
To avoid this, we recommend using `ignoreUrlPatterns` parameter when calling `enableHttpTracking` to ignore requests made by your Dio client.
```

brett0000FF commented on packages/datadog_dio/example/README.md:
```diff
@@ -0,0 +1,16 @@
+# example
```
```suggestion
# Example
```

brett0000FF reviewed (dismissed):
Left suggestions for a couple minor typo corrections.

fuzzybinary committed:
Fix Android integration tests

Needed to add clear text exceptions for the integration tests.

Address remaining code review comments.

fuzzybinary committed:
Apply suggestions from code review

Co-authored-by: Brett Blue <84536271+brett0000FF@users.noreply.github.com>

awforsythe commented on packages/datadog_dio/CHANGELOG.md:
```diff
@@ -0,0 +1,3 @@
+## 1.0.0
+
+
```
You may want to list something here? Even if it's just:

* Initial release of the Datadog Dio package
* Support for tracking HTTP requests via `addDatadogInterceptor`
* Custom attribute support via `DatadogDioAttributeProvider`

...or something similar.

fuzzybinary replied:
Good call. I'll add it.

awforsythe commented on packages/datadog_dio/lib/src/datadog_dio_interceptor.dart:
```diff
@@ -0,0 +1,192 @@
+      } catch (e, st) {
+        datadogSdk.internalLogger.sendToDatadog(
+          '$DatadogDioInterceptor encountered an error while attempting '
+          ' to track a request: $e',
```
nit: Extra space in error message (`while attempting  to track`)

awforsythe commented on packages/datadog_dio/lib/src/datadog_dio_interceptor.dart:
```diff
@@ -0,0 +1,192 @@
+      } catch (e, st) {
+        datadogSdk.internalLogger.sendToDatadog(
+          '$DatadogDioInterceptor encountered an error while attempting '
+          ' to track a request: $e',
```
nit: Extra space

awforsythe commented on packages/datadog_dio/test/datadog_dio_interceptor_test.dart:
```diff
@@ -0,0 +1,611 @@
+      verify(() => handler.next(requestOptions));
+    });
+
+    test('the interceptor calls attribute provider on request', () {
+      // Given
+      final attributeProvider = DatadogDioAttributeProviderMock();
+      when(() => attributeProvider.onRequest(any())).thenReturn(null);
+      final interceptor = DatadogDioInterceptor(
+          datadogSdk: mockDatadog, attributesProvider: attributeProvider);
+      final requestOptions =
+          RequestOptions(path: 'https://test_uri', method: 'POST', headers: {});
+
+      // When
+      final handler = RequestInterceptionHandlerMock();
+      interceptor.onRequest(requestOptions, handler);
+
+      // Then
+      verify(() => attributeProvider.onRequest(requestOptions));
+    });
```
nit: It might also be nice to verify that the interceptor handles exceptions thrown by user-provided `DatadogDioAttributeProvider` callbacks as expected

awforsythe commented on packages/datadog_dio/example/integration_test/dio_interceptor_test.dart:
```diff
@@ -0,0 +1,183 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2019-2022 Datadog, Inc.
```
nit: Copyright year should probably be updated? Looks like it's outdated (or the notice is missing) in several files within `packages/datadog_dio/example/`

fuzzybinary replied:
hrm, good call. I have a macro that I usually use to add the license. I'm not sure why this one is incorrect.

awforsythe reviewed (dismissed):
Looks great overall. I'd suggest adding something to the `CHANGELOG` for 1.0.0; everything else is a nitpick.

fuzzybinary committed:
Apply suggestions from code review. Fix copyrights

fuzzybinary commented:
Fixed most of the licenses in the `.dart` files. Not going to worry too much about the platform specific files in `example` though.
